test-graphql/client-test.js

```diff
@@ -0,0 +1,217 @@
+import assert from 'assert';
+import GraphQLJSClient from '@shopify/graphql-js-client';
+import Config from '../src-graphql/config';
+import Client from '../src-graphql/client';
+import types from '../types';
+import base64 from '../src-graphql/base64';
+import {singleProductFixture, multipleProductsFixture} from '../fixtures/product-fixture';
+import {fetchMock} from './fetch-mock-node';
+
+suite('client-test', () => {
+  const querySplitter = /[\s,]+/;
+
+  function tokens(queryString) {
+    return queryString.split(querySplitter).filter((token) => Boolean(token));
+  }
+
+  test('it instantiates a GraphQL client with the given config', () => {
+    let passedTypeBundle;
+    let passedUrl;
+    let passedFetchOptions;
+
+    class FakeGraphQLJSClient {
+      constructor(typeBundle, url, fetchOptions) {
+        passedTypeBundle = typeBundle;
+        passedUrl = url;
+        passedFetchOptions = fetchOptions;
+      }
+    }
+
+    const config = new Config({
+      domain: 'sendmecats.myshopify.com',
+      storefrontAccessToken: 'abc123'
+    });
+
+    new Client(config, FakeGraphQLJSClient);  // eslint-disable-line no-new
+
+    assert.equal(passedTypeBundle, types);
+    assert.equal(passedUrl, 'https://sendmecats.myshopify.com/api/graphql');
+    assert.deepEqual(passedFetchOptions, {
+      headers: {
+        'Content-Type': 'application/json',
+        Accept: 'application/json',
+        Authorization: `Basic ${base64(config.storefrontAccessToken)}`
+      }
+    });
+  });
+
+  test('it creates an instance of the GraphQLJSClient by default', () => {
+    const config = new Config({
+      domain: 'sendmecats.myshopify.com',
+      storefrontAccessToken: 'abc123'
+    });
+
+    const client = new Client(config);
+
+    assert.ok(GraphQLJSClient.prototype.isPrototypeOf(client.graphQLClient));
+  });
+
+  test('it resolves with an array of products on Client#fetchAll', () => {
+    const config = new Config({
+      domain: 'multiple-products.myshopify.com',
+      storefrontAccessToken: 'abc123'
+    });
+
+    const client = new Client(config);
+
+    fetchMock.post('https://multiple-products.myshopify.com/api/graphql', multipleProductsFixture);
+
+    return client.fetchAllProducts().then((products) => {
+      assert.ok(Array.isArray(products), 'products is an array');
+      assert.equal(products.length, 2, 'there are two products');
+
+      const passedArgs = fetchMock.lastCall('https://multiple-products.myshopify.com/api/graphql');
```

Comment:
It's not a side effect in this context since this isn't SDK usage, it's `fetchMock` usage. In this case it returns an array and we may not care about the first item (the URL). Destructuring makes the code simpler/more terse (to me at least) and that rule enables it.

Agreed no-unused-vars is a good thing for app code, but it's super annoying to selectively disable throughout tests. Any war to have different rules for app vs test code (which should generally be avoided but there are a few things like this).